RFC: s3 locking based on conditional writes (#2511)
opentofu/opentofu: rfc/20250211-s3-locking-with-conditional-writes.md

# S3 Backend: Locking by using the recently released feature of conditional writes

Issue: https://github.com/opentofu/opentofu/issues/599

> Considering the request from the ticket above, and the newly AWS released S3 feature, we can now have state locking without relying on another instance of DynamoDB.

Reviewer: remove "another instance of"

The main reasons for such a change could be summarized as follows:
* Less resources to maintain.

> * Removing costs that now are totally useless.

Reviewer: * Potentially reducing costs by eliminating usage of dynamo db.

Author: Thanks. Fixed 711eaf70265ee42a67c97c178b4cb21326460354

* One less point of failure (removing DynamoDB from the state management)

The most important things that need to be handled during this implementation:
* A simple way to enable/disable locking by using S3 conditional writes.
* A path forward to migrate the locking and state digest from DynamoDB to S3.
* The default behavior should be untouched as long as the `backend` block configuration contains no attributes related to the new locking option.

## Proposed Solution

Until recently, most of the approaches that could have been taken for this implementation, could have been prone to data races.

> But AWS released a new functionality to the S3, and now it supports conditional writes on any given S3 bucket.

Author: Thanks. Fixed 711eaf70265ee42a67c97c178b4cb21326460354

> The reason is to be able to check the integrity of the state object from the S3 bucket in a future run. This is done by reading the digest from DynamoDB and comparing it with the ETag attribute of the state object from S3. 

By moving to the S3 based locking, OpenTofu will hold no other file for the digest of the state object, since the digest is kept in `ETag` header of the object.
More info about this topic can be found on the [official documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html).

But if both locks are enabled (`use_lockfile=true` and `dynamodb_table=<actual_table_name>`), the digest file will still be stored in DynamoDB.

> [!WARNING]
> 
> By enabling the S3 locking and disabling the DynamoDB one, the digest from DynamoDB will become stale. This means that if it is desired to go back to the DynamoDB locking, the digest needs to be cleaned up or updated in order to allow the content integrity check to work.

### Open Questions

* Do we want to provide the option to have the lock objects into another bucket? This will break the feature parity.

### Future Considerations

> If this feature will prove to have a high adoption rate, later, we might consider to deprecate the DynamoDB locking mechanism. 

Reviewer 2: This reads as adoption rate by the OpenTofu users.
When in actuality to deprecate the DynamoDB locking we would need the adoption of the locking feature by the majority of S3 compatible storage services.
We should highlight that explicitly.

Author: Good point! Thanks for the input. Fixed in 12adf6200ff60dfe12597285173e08990cf5b3cb

* When the new attribute `use_lockfile` will exist **alongside** `dynamodb_table`, OpenTofu will:
  * Acquire the lock in the S3 bucket;
  * Acquire the lock in DynamoDB table;
  * Get the digest of the state object from DynamoDB and ensure the state object content integrity;
  * Perform the requested sub-command;
  * Release the lock from the S3 bucket;
  * Release the lock from DynamoDB table.

The usage of [workspaces](https://opentofu.org/docs/language/state/workspaces/) will not impact this new way of locking. The locking object will always be stored right next to its related state object.

> [!NOTE]
>
> OpenTofu [recommends](https://opentofu.org/docs/language/settings/backends/s3/) to have versioning enabled for the S3 buckets used to store state objects.
>
> Acquiring and releasing locks will add a good amount of writes and reads to the bucket. Therefore, for a versioning-enabled S3 bucket, the number of versions for that object could grow significantly.
> Even though the cost should be negligible for the locking objects, any user using this feature could consider configuring the lifecycle of the S3 bucket to limit the number of versions of an object.

> [!WARNING]
> 

> > When OpenTofu S3 backend is used with an S3 compatible provider, it needs to be checked that the provider is supporting conditional writes in the same way AWS S3 is offering.

Author: Applied in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179

#### Migration paths
##### I have no locking enabled
In this case, the user can just add the new `use_lockfile=true` and run `tofu init -reconfigure`.

##### I have DynamoDB locking enabled

> In case the user is having DynamoDB enabled, there are two paths forward:

Reviewer 3: ```suggestion
In case the user has DynamoDB enabled, there are two paths forward:
```

Author: Applied in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179

1. Add the new attribute `use_lockfile=true` and run `tofu init -reconfigure`

>    * Later, remove the `dynamodb_table` attribute and run `tofu init -reconfigure` again

Reviewer 3: > Later, remove the `dynamodb_table` attribute and run `tofu init -reconfigure` again

Not sure if I understand what "later" exactly means here. It may be also unclear for end-users. Could you please elaborate more on that one?

Author: Added more details in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179.
Please let me know if that is making clearer what the pros and cons of each is are.

> 2. Add the new attribute `use_lockfile=true`, remove the `dynamodb_table` one and run `tofu init -reconfigure`

Reviewer 3: Also, I am not sure what is the difference between these two approaches. I would prefer to stick to a single one or describe more when using the first or the second one is better.

Author: Added more details in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179.
Please let me know if that is making clearer what the pros and cons of each is are.

> OpenTofu recommends to have both locking mechanisms enabled for a limited amount of time and later remove the DynamoDB locking.

Reviewer 3: The same question as above, why do we recommend having both locking mechanisms enabled for a limited amount of time?

Author: Added more details in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179.
Please let me know if that is making clearer what the pros and cons of each is are.

* HTTP 200 (OK) - Means that the locking object was not existing. Therefore, the lock can be considered as acquired.
  * For buckets with versioning enabled, if there's no current object version with the same name, or if the current object version is a delete marker, the write operation succeeds.
* HTTP 412 (Precondition Failed) - Means that the locking object is already there. Therefore, the whole process should exit because the lock couldn't be acquired.
* HTTP 409 (Conflict) - Means that the there was a conflict of concurrent requests. AWS recommends to retry the request in such cases, but we could just handle this similarly to the 412 case.

#### Digest updates
> [!NOTE]
> Right now, when locking is enabled on DynamoDB, at the moment of updating the state object content, OpenTofu also writes an entry in DynamoDB with the MD5 sum of the state object.
> The reason is to be able to check the integrity of the state object from the S3 bucket in a future run. This is done by reading the digest from DynamoDB and comparing it with the ETag attribute of the state object from S3. 

> By moving to the S3 based locking, OpenTofu will hold no other file for the digest of the state object, since the digest is kept in `ETag` header of the object.

Reviewer 3: > since the digest is kept in `ETag` header of the object

I believe digest check was introduced to tackle the eventual consistency nature of s3 before 2020, so we can clearly state this is deprecated behavior and we don't need that at all. Unless, we also want to check if that object hasn't been updated, but that part we handle via a separate lock file. So I am not sure if checksum validation is something we want to keep.

WDYT?

Author: Reworked in 6c43f94d46a40ebf22b652cb08d42bd5ba7f4179.
Yes, this is the idea. Once only the S3 locking is in use, the digest of the state will not be handled anymore by OpenTofu.

More info about this topic can be found on the [official documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html).

But if both locks are enabled (`use_lockfile=true` and `dynamodb_table=<actual_table_name>`), the digest file will still be stored in DynamoDB.

> [!WARNING]
> 
> By enabling the S3 locking and disabling the DynamoDB one, the digest from DynamoDB will become stale. This means that if it is desired to go back to the DynamoDB locking, the digest needs to be cleaned up or updated in order to allow the content integrity check to work.

### Open Questions

* Do we want to provide the option to have the lock objects into another bucket? This will break the feature parity.

### Future Considerations
Later, DynamoDB based locking might be considered for deprecation once (at least) the following are true:

> * Conditional writes are implemented in the majority of the S3 compatible services.

Reviewer 3: That's an interesting question. How do people handle locking with non-AWS S3 compatible storage? Do they use DynamoDB? I believe it is easier to switch to AWS S3 if you are already in use with DynamoDB service

Author: This would be a question for the community TBH.

Reviewer 3: Yep, agree. Just wanted to share a thought

Reviewer: My understanding is that the only real option is to run with locking disabled.  AFAIK there's no way to separate the credentials for S3 from the credentials from Dynamo

# S3 Backend: Locking by using the recently released feature of conditional writes

Issue: https://github.com/opentofu/opentofu/issues/599

Considering the request from the ticket above, and the newly AWS released S3 feature, we can now have state locking without relying on DynamoDB.

The main reasons for such a change could be summarized as follows:
* Less resources to maintain.
* Potentially reducing costs by eliminating usage of dynamo db.

> * One less point of failure (removing DynamoDB from the state management)

Reviewer: Additonal point:
* Easy for other S3 compatible services to implement and enable locking.

Author: Added in a3139cf174b05fdcd85dab98b3726874d9a10c4f

##### I have DynamoDB locking enabled
In case the user has DynamoDB enabled, there are two paths forward:
1. Add the new attribute `use_lockfile=true` and run `tofu init -reconfigure`
   * Later, after a baking period with both locking mechanisms enabled, if no issues encountered, remove the `dynamodb_table` attribute and run `tofu init -reconfigure` again.
   * Letting both locking mechanisms enabled, ensures that nobody will acquire the lock regardless of having or not the latest configuration.
2. Add the new attribute `use_lockfile=true`, remove the `dynamodb_table` one and run `tofu init -reconfigure` 
   * **Caution:** when the configuration updated is executed from multiple places (multiple machines, pipelines on PRs, etc), you might get into issues where one outdated copy of the configuration is using DynamoDB locking and the one updated is using S3 locking. This could create concurrent access on the same state.
   * Once the state is updated by using this approach, the state digest that OpenTofu was storing in DynamoDB (for data consistency checks) will get stale. If it is wanted to go back to DynamoDB locking, the old digest needs to be cleaned up manually.

> OpenTofu recommends to have both locking mechanisms enabled for a limited amount of time and later remove the DynamoDB locking. This is to ensure that the changes pushed upstream will be propagated into all of your places that the configuration could be executed from.

Reviewer 4: Ok, now I understand the reason :)

# S3 Backend: Locking by using the recently released feature of conditional writes

Issue: https://github.com/opentofu/opentofu/issues/599

Considering the request from the ticket above, and the newly AWS released S3 feature, we can now have state locking without relying on DynamoDB.

The main reasons for such a change could be summarized as follows:
* Less resources to maintain.
* Potentially reducing costs by eliminating usage of dynamo db.
* One less point of failure (removing DynamoDB from the state management)

> The most important things that need to be handled during this implementation:

Author: Really good point!
From a quick check, I see that some already implemented it but there are still some that didn't yet.
Added a warning about it e22385a328c24a07904fa422aad15849b57ff047

> The most important things that need to be handled during this implementation:

Reviewer 3: We need not to forget to update the user-facing docs as well.

> [!NOTE]
>
> The name `use_lockfile` was selected this way to keep the [feature parity with Terraform](https://developer.hashicorp.com/terraform/language/backend/s3#state-locking).

When OpenTofu will find the above attribute in the backend configuration, it will handle the state locking and the digest of the state file in the same bucket as configured for the state file.

### User Documentation

In order to make use of this new feature, the users will have to add the attribute in the `backend` block:
```terraform
terraform {
  backend "s3" {
    bucket = "tofu-state-backend"
    key = "statefile"
    region = "us-east-1"
    dynamodb_table = "tofu_locking"
    use_lockfile = true
  }
}
```

* When the new attribute `use_lockfile` exists and `dynamodb_table` is missing, OpenTofu will behave as normal:
  * The only thing that will have to do extra, is to ensure that the MD5 file is there and if not, to create it.
    * The generation of the MD5 checksum is already part of the [implementation](https://github.com/opentofu/opentofu/blob/6614782/internal/backend/remote-state/s3/client.go#L178).

>     * We just need to store it in a new file in the configured S3 bucket (file name: <state-file-path>-md5)

Author: Checked. [This](https://github.com/opentofu/opentofu/blob/6614782e6da121c1e1464a9471d943c7cc939a6b/internal/backend/remote-state/s3/backend_state.go#L229-L235) part is handling that.
Also, in the [OpenTofu S3 docs](https://opentofu.org/docs/language/settings/backends/s3/), that's clearly stated:
![image](https://github.com/user-attachments/assets/6259799c-211b-42e9-a8df-d61e4fb592c5)

>     * We just need to store it in a new file in the configured S3 bucket (file name: <state-file-path>-md5)

Reviewer 3: Yep, I meant we should add a comment to the RFC stating that it will not cause any problem when using multiple workspaces.

>     * We just need to store it in a new file in the configured S3 bucket (file name: <state-file-path>-md5)

Author: Added in 8ff0b16733c70782c209c3f5a023190a35f93bab

* When the new attribute `use_lockfile` will exist **alongside** `dynamodb_table`, OpenTofu will:
  * Acquire the lock in the S3 bucket;
  * Get the digest of the statefile from S3. If missing:
    * Acquire the lock in DynamoDB table; 
    * Get the digest of the statefile from DynamoDB and migrate it to the S3 bucket;
    * Release the lock from DynamoDB table;
  * Perform the requested sub-command

>   * Release the lock from the S3 bucket;

Author: Thanks. Added 7b88c6f84ddb90024fff88eec0fa7f8f3fa7ad6f

But the simplified implementation would look like this:
```go
input := &s3.PutObjectInput{
    Bucket:            aws.String(bucket),
    Key:               aws.String(key),
    Body:              bytes.NewReader([]byte(lockInfo)),
    IfNoneMatch:       aws.String("*"),
}
_, err := actor.S3Client.PutObject(ctx, input)
```

The `err` returned above should be handled accordingly with the [behaviour defined](https://docs.aws.amazon.com/AmazonS3/latest/userguide/conditional-writes.html) in the official docs:
* HTTP 200 (OK) - Means that the locking file was not existing. Therefore, the lock can be considered as acquired.
  * For buckets with versioning enabled, if there's no current object version with the same name, or if the current object version is a delete marker, the write operation succeeds.
* HTTP 412 (Precondition Failed) - Means that the locking file is already there. Therefore, the whole process should exit because the lock couldn't be acquired.
* HTTP 409 (Conflict) - Means that the there was a conflict of concurrent requests. AWS recommends to retry the request in such cases, but we could just handle this similarly to the 412 case.

> #### Digest updates

Reviewer 3: I may be missing something, but do we need to update the digest? If I understand correctly, we could go with a `IfNoneMatch` lock creation so it would fail for any other tofu run, while the file exists. On Unlock, we could just delete it altogether. As far as I can tell, this matches the dynamodb implementation and eliminates data races via `IfNoneMatch` option. 

What do you think?

Author: I used terms like "digest", "md5" or "digest file" for the part where OpenTofu is writing an md5 hash of the statefile.
For example, looking into the dynamoDB, there is the md5 sum of the state file. Can be easily checked like this:
![image](https://github.com/user-attachments/assets/7dc94a57-407b-4531-b6db-02bc2e24d10b)

Reviewer 3: Got you, thanks! I missed that part, I will re-read it and leave another comment.

Author: Updated this section due to not needed to handle this digest file anymore.
This should be handled only when the DynamoDB locking is enabled. That's due to having two sources of information when it comes to locking and it was used to ensure that the state object from the S3 bucket was not altered separately, without acquiring a lock.

Reviewer 3: After some digging, I think this option is unnecessary even for dynamodb locking. This feature was introduced to workaround the eventual consistency nature of s3 (before 2020). It is checking if the read state file is the latest written one and we no longer need to check this since s3 become strongly consistent ([blogpost](https://aws.amazon.com/blogs/aws/amazon-s3-update-strong-read-after-write-consistency/)) in 2020.

Reviewer 3: This is something only slightly relevant for this RFC, so I think we can tackle it separately. Alternatively, we can leave it as is if we plan to deprecate dynamodb locking completely. Otherwise, removing md5 write / read from dynamodb locking will positively affect the performance.

Author: Added in 38f1eb921775d9ac783fffe8c8cb3d2b11a67419

	err := putMD5(sum, "") // Empty ETag should rewrite the digest object
}
```

### Open Questions

* Do we want to provide the option to have the lock objects into another bucket? This will break the feature parity.
* When writing the digest (*-md5) object, should we consider conditional writing by using the `IfMatch` argument?
  ```go
  func (actor S3Actions) UpdateDigest(ctx context.Context, bucket string, key string, beforeApplyDigest string) error {
      newDigest := uuid.NewString()
      input := &s3.PutObjectInput{
          Bucket:            aws.String(bucket),
          Key:               aws.String(key),
          Body:              bytes.NewReader([]byte(newDigest)),
          ChecksumAlgorithm: types.ChecksumAlgorithmSha256,
          IfMatch:           aws.String(beforeApplyDigest),
      }
      _, err := actor.S3Client.PutObject(ctx, input)
      return err
  }
  ```
  Asking because this could also lead to other issues in case it doesn't match. I would suggest to have this only for showing an error in case the digest is different.
* Right now, since the MD5 digest is stored in DynamoDB, this is written only when the `dynamodb_table` is specified.

>   * Should we do the same in the new implementation? Write the MD5 digest only if the `use_lockfile=true`?

Reviewer 3: I think we should only write lock files if the locking is enabled, so I would say yes.

Author: Removed some questions since are not needed anymore. More details in the [above comment](https://github.com/opentofu/opentofu/pull/2511#discussion_r1955828819).
